ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BaseClusterRequest.java

```diff
@@ -224,4 +210,73 @@ public ProvisionAction getProvisionAction() {
   public void setProvisionAction(ProvisionAction provisionAction) {
     this.provisionAction = provisionAction;
   }
+
+  /**
+   * @return the request converted to a {@link TopologyRequestEntity}
+   */
+  public TopologyRequestEntity toEntity() {
+    TopologyRequestEntity entity = new TopologyRequestEntity();
+
+    //todo: this isn't set for a scaling operation because we had intended to allow multiple
+    //todo: bp's to be used to scale a cluster although this isn't currently supported by
+    //todo: new topology infrastructure
+    entity.setAction(getType().name());
+    if (getBlueprint() != null) {
+      entity.setBlueprintName(getBlueprint().getName());
+    }
+
+    entity.setClusterAttributes(JsonUtils.toJson(getConfiguration().getAttributes()));
+    entity.setClusterId(getClusterId());
+    entity.setClusterProperties(JsonUtils.toJson(getConfiguration().getProperties()));
+    entity.setDescription(getDescription());
+
+    if (getProvisionAction() != null) {
+      entity.setProvisionAction(getProvisionAction());
+    }
+
+    // host groups
+    Collection<TopologyHostGroupEntity> hostGroupEntities = new ArrayList<>();
+    for (HostGroupInfo groupInfo : getHostGroupInfo().values())  {
+      hostGroupEntities.add(toHostGroupEntity(groupInfo, entity));
+    }
+    entity.setTopologyHostGroupEntities(hostGroupEntities);
+
+    return entity;
+  }
+
+  private TopologyHostGroupEntity toHostGroupEntity(HostGroupInfo groupInfo, TopologyRequestEntity topologyRequestEntity) {
+    TopologyHostGroupEntity entity = new TopologyHostGroupEntity();
+    entity.setGroupAttributes(JsonUtils.toJson(groupInfo.getConfiguration().getAttributes()));
+    entity.setGroupProperties(JsonUtils.toJson(groupInfo.getConfiguration().getProperties()));
+    entity.setName(groupInfo.getHostGroupName());
+    entity.setTopologyRequestEntity(topologyRequestEntity);
+
+    // host info
+    Collection<TopologyHostInfoEntity> hostInfoEntities = new ArrayList<>();
+    entity.setTopologyHostInfoEntities(hostInfoEntities);
+
+    Collection<String> hosts = groupInfo.getHostNames();
+    if (hosts.isEmpty()) {
+      TopologyHostInfoEntity hostInfoEntity = new TopologyHostInfoEntity();
+      hostInfoEntity.setTopologyHostGroupEntity(entity);
+      hostInfoEntity.setHostCount(groupInfo.getRequestedHostCount());
+      if (groupInfo.getPredicate() != null) {
+        hostInfoEntity.setPredicate(groupInfo.getPredicateString());
+      }
+      hostInfoEntities.add(hostInfoEntity);
+    } else {
+      for (String hostName : hosts) {
+        TopologyHostInfoEntity hostInfoEntity = new TopologyHostInfoEntity();
+        hostInfoEntity.setTopologyHostGroupEntity(entity);
+        if (groupInfo.getPredicate() != null) {
```

Comment:
Is the if-check necessary? Can you just set it to the value of the predicate, even if the predicate is null?

Reply:
I actually didn't write this method but copied it over from PersistedStateImpl. I can change it nevertheless.